src/hotspot/share/opto/graphKit.hpp

```diff
@@ -275,7 +275,10 @@ class GraphKit : public Phase {
 
   // Helper to throw a built-in exception.
   // The JVMS must allow the bytecode to be re-executed via an uncommon trap.
-  void builtin_throw(Deoptimization::DeoptReason reason);
+  // If `exception_object` is nullptr, the exception to throw will be guessed based on `reason`
+  void builtin_throw(Deoptimization::DeoptReason reason, ciInstance* exception_object = nullptr);
```

Comment:
Please, introduce a new overload instead. 

I suggest to extract Deoptimization::DeoptReason -> ciInstance mapping into a helper method and turn `void builtin_throw(Deoptimization::DeoptReason reason)` into a wrapper:
```
void GraphKit::builtin_throw(Deoptimization::DeoptReason reason) {
   builtin_throw(reason, exception_on_deopt(reason));
}
```